onnxruntime/python/onnxruntime_pybind_sparse_tensor.cc

```diff
@@ -0,0 +1,347 @@
+          }
+
+          // go ahead and create references to make sure storage does not disappear
+          std::vector<py::object> reference_holders = {py_values, py_inner_indices, py_outer_indices};
+          OrtMemoryInfo mem_info = GetMemoryInfoPerDeviceType(ort_device);
+          auto sparse_tensor = std::make_unique<SparseTensor>(ml_type, dense_shape, mem_info);
+          ORT_THROW_IF_ERROR(sparse_tensor->RepBuilder<SparseCsrcBuilder>()
+                                 .Create(SparseCsrcFormatRep::kRowMajor,
+                                         py_values.size(), py_inner_indices.size(), py_outer_indices.size(),
+                                         const_cast<void*>(py_values.data()),
+                                         const_cast<int64_t*>(py_inner_indices.data()),
+                                         const_cast<int64_t*>(py_outer_indices.data())));
+          result = std::make_unique<PySparseTensor>(std::move(sparse_tensor), std::move(reference_holders));
+        } else if (values_type == NPY_UNICODE || values_type == NPY_STRING) {
+          if (ort_device.Type() != OrtDevice::CPU) {
+            throw std::runtime_error("Only CPU based devices are supported for non-numeric datatypes");
+          }
+          auto sparse_tensor = std::make_unique<SparseTensor>(ml_type, dense_shape, GetAllocator());
+          SparseCsrcFormatRep* rep;
+          ORT_THROW_IF_ERROR(sparse_tensor->RepBuilder<SparseCsrcBuilder>()
+                                 .Create(SparseCsrcFormatRep::kRowMajor,
+                                         py_values.size(), py_inner_indices.size(), py_outer_indices.size(),
+                                         rep));
+          CopyDataToTensor(py_values, values_type, rep->MutableValues());
+          CopyDataToTensor(py_inner_indices, GetNumpyArrayType(py_inner_indices), rep->MutableInner());
+          CopyDataToTensor(py_outer_indices, GetNumpyArrayType(py_outer_indices), rep->MutableOuter());
+          result = std::make_unique<PySparseTensor>(std::move(sparse_tensor));
+        } else {
+          ORT_THROW("Unsupported values data type: ", values_type);
+        }
+
+        return result;
+      })
+      // Returns a numpy array that is backed by SparseTensor values memory
+      // be aware that it may be on GPU
+      .def("values", [](const PySparseTensor* py_tensor) -> py::array {
+        const SparseTensor& sparse_tensor = py_tensor->Instance();
+        if (sparse_tensor.IsDataTypeString()) {
+          // Strings can not be on GPU but we can not expose them as mapped memory
+          // So we need to create a copy.
+          py::list str_list;
+          auto str_span = sparse_tensor.Values().DataAsSpan<std::string>();
+          for (const auto& s : str_span) {
+            // valid UTF-8 to python UNICODE is handled automatically
+            str_list.append(py::cast(s));
+          }
+          return py::cast<py::array>(str_list);
```

Comment:
>:cast<py::array [](http://example.com/codeflow?start=20&length=15)

Need to preserve shape in the future, when values have complex shape with block sparse format